### ISSUE TITLE
Allow files to be excluded from organizeImports

### ISSUE THREAD

**Issue by dshinton-qstart** [microsoft/TypeScript#54090]
# Suggestion

## 🔍 Search Terms

* organizeImports

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I use `source.organizeImports` as an automated action on save in VS Code. It works great for 99% of my files. However, for some files, it will break my application due to it being destructive. For the 1% of files it will break, I would like `source.organizeImports` to ignore them.

Would it be possible to add a feature where there is a way to exclude certain files from `source.organizeImports` to prevent it from running on files it will break?

Thank you!

## 📃 Motivating Example

In my web application, I have a `main.js` file that has imports that are not directly used, but needed by other files. The `organizeImports` feature removes these imports, which breaks my application.

## 💻 Use Cases

The only workaround for this is to either

* Not automatically run `organizeImports` on save
* Not edit files that `organizeImports` will break using VS Code
* Use `sortImports` and lose the automated cleanup `organizeImports` provides

None of these workarounds are ideal.

**Comment by fatcerberus** [comment]
`importsNotUsedAsValues: "preserve"`?

**Comment by dshinton-qstart** [comment]
> `importsNotUsedAsValues: "preserve"`

Perhaps I am not using it properly, but adding that directive to the `compilerOptions` of my `jsconfig.json` or `tsconfig.json` has no effect.

> Although I’m confused why this breaks you at runtime

The way Font Awesome icons are selectively loaded in Vue looks like:

```
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import library from './fontawesome-icons.js'
```

That second import is not referenced in the main.js file, but is needed to load icons into the application. `organizeImports` removes that import since it thinks it is unused.

**Comment by fatcerberus** [comment]
> That second import is not referenced in the main.js file, but is needed to load icons into the application.

Yeah, I got that, I’m just confused why *TypeScript itself* isn’t removing the import at compile time.  Usually it would.

**Comment by fatcerberus** [comment]
btw if you’re intentionally importing *only* for side effects with no intent to use the imported value(s) then you can just write `import "foo";` and that should prevent the import from being removed.
